### PR TITLE
Alert Module

### PR DESCRIPTION
**Pull request by martunjai**
Alert Module

### CODE REVIEW

File: ReactSkia/sdk/WindowDelegator.cpp [hunk 1]
```diff
@@ -30,19 +30,21 @@ void WindowDelegator::createWindow(SkSize windowSize,std::function<void ()> wind
...
 
   if(displayPlatForm_ == RnsShell::PlatformDisplay::Type::X11) {
     /*For X11 draw should be done after expose event received*/
     sem_init(&semReadyToDraw_,0,0);
     // Registering expose event
-    std::function<void(RnsShell::Window*)> handler = std::bind(&WindowDelegator::onExposeHandler,this,
-                                                                         std::placeholders::_1);
+    std::function<void(RnsShell::Window*)> handler = std::bind(&WindowDelegator::onExposeHandler, this, std::placeholders::_1);
     exposeEventID_ = NotificationCenter::defaultCenter().addListener("windowExposed",handler);
+    RNS_LOG_INFO("INSIDE::createNativeWindow::1::exposeEventID_ = " << exposeEventID_);
```
Comment (martunjai) [inline ReactSkia/sdk/WindowDelegator.cpp]: remove Log if not required

File: rns_shell/platform/graphics/x11/WindowX11.cpp [hunk 2]
```diff
@@ -47,6 +49,7 @@ Window* Window::createNativeWindow(void* platformData,SkSize dimension,WindowTyp
         mainWindow_ = window;
     }
     window->winType=type;
+    i++;
```
Comment (martunjai) [inline rns_shell/platform/graphics/x11/WindowX11.cpp]: not required remove

File: rns_shell/platform/graphics/x11/WindowX11.cpp [hunk 3]
```diff
@@ -22,6 +22,8 @@
 
 #include "x11/PlatformDisplayX11.h"
 
+static int i = 0;
```
Comment (martunjai) [inline rns_shell/platform/graphics/x11/WindowX11.cpp]: not required remove

File: rns_shell/platform/graphics/libwpe/WindowLibWPE.cpp [hunk 4]
```diff
@@ -265,11 +265,11 @@ void WindowLibWPE::setRequestedDisplayParams(const DisplayParams& params, bool a
...
 
 void WindowLibWPE::onKey(rnsKey eventKeyType, rnsKeyAction eventKeyAction){
-#if ENABLE(FEATURE_ONSCREEN_KEYBOARD)
+// #if ENABLE(FEATURE_ONSCREEN_KEYBOARD)
     if(winType == SubWindow)
-       NotificationCenter::subWindowCenter().emit("onHWKeyEvent", eventKeyType, eventKeyAction);
+       NotificationCenter::subWindowCenter().emit("onHWKeyEvent", eventKeyType, eventKeyAction, (Window*)this);
     else
-#endif/*FEATURE_ONSCREEN_KEYBOARD*/
+// #endif/*FEATURE_ONSCREEN_KEYBOARD*/
```
Comment (martunjai) [inline rns_shell/platform/graphics/libwpe/WindowLibWPE.cpp]: uncomment this

File: rns_shell/platform/graphics/x11/WindowX11.cpp [hunk 5]
```diff
@@ -345,12 +345,12 @@ void WindowX11::onExpose() {
     NotificationCenter::defaultCenter().emit("windowExposed",(Window*)this);
 }
 
-void WindowX11::onKey(rnsKey eventKeyType, rnsKeyAction eventKeyAction){
-#if ENABLE(FEATURE_ONSCREEN_KEYBOARD)
+void WindowX11::onKey(rnsKey eventKeyType, rnsKeyAction eventKeyAction) {
+// #if ENABLE(FEATURE_ONSCREEN_KEYBOARD)
```
Comment (martunjai) [inline rns_shell/platform/graphics/x11/WindowX11.cpp]: uncomment this

File: ReactSkia/sdk/WindowDelegator.cpp [hunk 6]
```diff
@@ -30,19 +30,20 @@ void WindowDelegator::createWindow(SkSize windowSize,std::function<void ()> wind
 
 void  WindowDelegator::createNativeWindow() {
 
-  displayPlatForm_=RnsShell::PlatformDisplay::sharedDisplayForCompositing().type();
+  displayPlatForm_ = RnsShell::PlatformDisplay::sharedDisplayForCompositing().type();
```
Comment (shefayaparwin) [inline ReactSkia/sdk/WindowDelegator.cpp:58]: WindowDelegator.cpp has only changes w.r.t spacing. If it Is a intentional changes, then fine, else please do revert it back

Comment (martunjai) [reply ReactSkia/sdk/WindowDelegator.cpp:58]: this is an intentional change.

File: ReactSkia/core_modules/RSkAlertManager.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+#include "core_modules/Alert.h"
+#include "include/core/SkFontMgr.h"
+#include "include/core/SkCanvas.h"
+#include "include/core/SkColor.h"
+#include "include/core/SkFont.h"
+#include "sdk/NotificationCenter.h"
+#include "sdk/RNSKeyCodeMapping.h"
+#include "sdk/WindowDelegator.h"
+#include "ReactCommon/TurboModule.h"
+#include "ReactSkia/JSITurboModuleManager.h"
```
Comment (pprabhu-ng) [inline ReactSkia/core_modules/RSkAlertManager.h:19]: header file includes to follow coding guidelines and to be added uniformly
ex: 
sdk/NotificationCenter.h  and ReactSkia/JSITurboModuleManager.h 
one is having "ReactSkia" path other one is not having

File: ReactSkia/core_modules/RSkAlertManager.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace facebook {
+namespace react {
+
+enum AlertWindowState {
```
Comment (pprabhu-ng) [inline ReactSkia/core_modules/RSkAlertManager.h:29]: indentation correction

File: ReactSkia/core_modules/RSkAlertManager.h [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        ALERT_WINDOW_CREATE = 1, // Window is Created
+        ALERT_WINDOW_ACTIVE = 2, // Window is Active
+        ALERT_WINDOW_INACTIVE = 3, // Window is In-Active
+        ALERT_WINDOW_DESTRUCT = 4 // Window is Destructed
+    };
+
+class RSkAlertManager : public rns::sdk::WindowDelegator, public facebook::react::TurboModule {
+public:
+
+    RSkAlertManager(
```
Comment (pprabhu-ng) [inline ReactSkia/core_modules/RSkAlertManager.h:39]: this class is already under facebook::react namespace , so do we need to use it during referencing

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+facebook::jsi::Value RSkAlertManager::alertWithArgsWrapper(
+    facebook::jsi::Runtime &rt,
+    TurboModule &turboModule,
+    const facebook::jsi::Value *args,
+    size_t count) {
+  auto &alertManagerInstance = static_cast<RSkAlertManager &>(turboModule);
+  auto alertProps = facebook::jsi::dynamicFromValue(rt, args[0]);
+  alertManagerInstance.processAlertMessages(std::make_shared<Alert>(
+      alertProps["title"].getString(), alertProps["message"].getString()));
```
Comment (pprabhu-ng) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: "message" object can be either message string or undefined as per definition in Alert.js
So should we check before calling getString()

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    WindowDelegator::windowDelegatorCanvas->drawSimpleText(
+        alertRef->getTitle().c_str(),
+        strlen(alertRef->getTitle().c_str()),
+        SkTextEncoding::kUTF8,
+        drawStartPointX,
+        drawStartPointY,
+        font_,
+        paint_);
+
+    drawStartPointY = ((mainWindowSize_.height() - FONT_SIZE) / 2) + (FONT_SIZE) + OFFSET_VERTICAL;
```
Comment (pprabhu-ng) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: should we check if message is there or no before drawing ?

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+ */
+
+#include "RSkAlertManager.h"
+
+#include "sdk/sdkStylesConfig.h"
+
+namespace facebook {
+namespace react {
+
+std::mutex lockMsgDrawing;
```
Comment (pprabhu-ng) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: should this be global or part of RSkAlertManager ?

File: ReactSkia/core_modules/Alert.h [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+#pragma once
+
+namespace facebook {
+namespace react {
+
+class Alert {
```
Comment (pprabhu-ng) [inline ReactSkia/core_modules/Alert.h:13]: why is this class defined seperately ?

File: ReactSkia/core_modules/RSkAlertManager.cpp (hunk 12 above)
Comment (martunjai) [reply ReactSkia/core_modules/RSkAlertManager.cpp]: Currently, we are keeping this as part of RSkALertManager only.

File: ReactSkia/core_modules/Alert.h (hunk 13 above)
Comment (martunjai) [reply ReactSkia/core_modules/Alert.h:13]: We want to have a separate Alert class where we can keep alert-related Parameters in the same place and use it accordingly.

For Eg.:
title,
message,
callbackOrButtons,
type,
defaultValue,
keyboardType,
options

File: ReactSkia/core_modules/RSkAlertManager.cpp (hunk 11 above)
Comment (martunjai) [reply ReactSkia/core_modules/RSkAlertManager.cpp]: yes, we should be checking if there is a message or not.

File: ReactSkia/core_modules/RSkAlertManager.cpp (hunk 10 above)
Comment (martunjai) [reply ReactSkia/core_modules/RSkAlertManager.cpp]: Based on the current implementation,
two arguments are required for Alert Constructor.

While drawing messages on Alert Box, we are keeping a check if the message string is null or not,
then based on the decision we will draw a message.

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+  displayMsgIndex_ = alertPropsList_.size();
+}
+
+void RSkAlertManager::windowReadyToDrawCB() {
+  alertWindowState_ = ALERT_WINDOW_ACTIVE;
+  drawAlertMsg();
+  commitDrawCall();
+}
+
+void RSkAlertManager::createAlertWindow() {
```
Comment (martunjai) [inline ReactSkia/core_modules/RSkAlertManager.cpp:129]: create and draw alert

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+#include "ReactSkia/core_modules/RSkAlertManager.h"
+#include "ReactSkia/sdk/sdkStylesConfig.h"
+
+namespace facebook {
+namespace react {
+
+std::mutex lockMsgDrawing;
+
+#define FONT_SIZE 24
```
Comment (martunjai) [inline ReactSkia/core_modules/RSkAlertManager.cpp:16]: TODO:
Will have to use different values for 720p and 1080p

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+  setWindowTittle("Alert Window!!!");
+  createWindow(mainWindowSize_, createWindowCB, forceFullScreenDrawCB);
+  if (subWindowKeyEventId_ == -1) {
+    std::function<void(rnsKey, rnsKeyAction, RnsShell::Window*)> handler =
+        std::bind(
+            &RSkAlertManager::onHWKeyHandler,
+            this,
+            std::placeholders::_1,  // KeyValue
+            std::placeholders::_2,  // eventKeyAction
+            std::placeholders::_3); // Window
```
Comment (martunjai) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: TODO:
Add something

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+RSkAlertManager::RSkAlertManager(
+    const std::string &name,
+    std::shared_ptr<facebook::react::CallInvoker> jsInvoker,
+    facebook::react::Instance *bridgeInstance)
+    : TurboModule(name, jsInvoker) {
+  methodMap_["alertWithArgs"] = MethodMetadata{1, alertWithArgsWrapper};
+}
+
+void RSkAlertManager::processAlertMessages(std::shared_ptr<Alert> alertPtr) {
+  alertPropsList_.push_back(alertPtr);
```
Comment (martunjai) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: TODO: Change Name

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+        strlen(alertRef->getTitle().c_str()),
+        SkTextEncoding::kUTF8,
+        drawStartPointX,
+        drawStartPointY,
+        font_,
+        paint_);
+
+
+    drawStartPointY = ((mainWindowSize_.height() - FONT_SIZE) / 2) + (FONT_SIZE) + OFFSET_VERTICAL;
+    if (alertRef->getMessage().c_str() != NULL) {
```
Comment (munezbn) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: Check if string is empty instead.

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+      alertRef->getTitle().c_str(),
+      strlen(alertRef->getTitle().c_str()),
+      SkTextEncoding::kUTF8);
+
+  if (!alertPropsList_.empty()) {
+    drawStartPointX = (mainWindowSize_.width() - titleWidth) / 2;
+    drawStartPointY = (mainWindowSize_.height() - FONT_SIZE) / 2;
+    WindowDelegator::windowDelegatorCanvas->drawSimpleText(
+        alertRef->getTitle().c_str(),
+        strlen(alertRef->getTitle().c_str()),
```
Comment (munezbn) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: Use built in string function to get length.

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+        SkTextEncoding::kUTF8,
+        drawStartPointX,
+        drawStartPointY,
+        font_,
+        paint_);
+
+
+    drawStartPointY = ((mainWindowSize_.height() - FONT_SIZE) / 2) + (FONT_SIZE) + OFFSET_VERTICAL;
```
Comment (munezbn) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: Instead of overwriting , append using +=

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+  }
+  font_.setSize(FONT_SIZE);
+  paint_.setColor(DARK_THEME_FONT_COLOR);
+}
+
+void RSkAlertManager::onHWKeyHandler(
+    rnsKey keyValue,
+    rnsKeyAction eventKeyAction,
+    RnsShell::Window *window) {
+  if ((eventKeyAction != RNS_KEY_Press) || (getWindow() != window)) {
```
Comment (munezbn) [inline ReactSkia/core_modules/RSkAlertManager.cpp:155]: Window check is a temp workaround. So add a TODO comment here. We should avoid passing entire window and try for ID alone.

File: ReactSkia/core_modules/RSkAlertManager.cpp [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+    return;
+  }
+
+  RNS_LOG_INFO("KEY RECEIVED : " << RNSKeyMap[keyValue]);
+
```
Comment (munezbn) [inline ReactSkia/core_modules/RSkAlertManager.cpp]: This is not thread safe. We need to lock in push pop and draw calls.

**Comment by shefayaparwin** [comment]
Closing This PR. It is replaced with below
https://github.com/nagra-opentv/react-native-skia/pull/163

**Pull request closed without merging**